### PR TITLE
feat(apps): allow loading apps from subdirectory

### PR DESCRIPTION
**Pull request by pjhul**
## Problem

Currently, you're unable to load an app from GitHub that isn't located at the root directory. As part of the goal of centralizing all our apps in a monorepo, we'll need this ability.

## Changes

- Allows passing a GitHub URL with an optional path (eg `https://github.com/pjhul/posthog-test-app/tree/main/app`)
- Extracts all files from this path 
- Fixes a bug where loading a plugin at a specific tag/branch/release would get overridden to the latest commit on main when upgrading
   - This is done by storing the full GitHub URL on each plugin rather than just the root (`https://github.com/pjhul/posthog-test-app`)

## Future work

- [ ] When loading a plugin from a branch instead of a specific commit, the way we check for updates will not fetch the most recent commit on this branch (Will fix in a future PR)
- [ ] We don't have any unit tests specifically for `extract_plugin_code`

## How did you test this code?

- Installed a plugin locally from a repo that was located in a subdirectory (`https://github.com/pjhul/posthog-test-app/tree/main/app`)
- Confirmed installing a plugin without a path still succeeds
- Added unit tests for parsing GitHub URLs
- Adds a new sample to `plugin_archives` to test loading an app from a subdirectory

### CODE REVIEW

**Comment by mariusandra** [comment]
This looks good, though I didn't run the code nor give it a deep code review / QA session. 

The main thing I'd double check is if upgrading packages works well. E.g. if two packages are installed from a subfolder and you click "check for updates", will they both show as "to update", even if the files changed for just one of them, or not. It could work really well, I just didn't try :).

However I also wanted to point out that if we're going for a monorepo, there's one more way to solve what this PR is trying to solve: npm packages. 

We could keep all apps inside one repo, and publish them individually as npm packages after each change. Then we'd just use these NPM urls instead of Github urls in the repository json. Probably not worth it if this PR lands, but just thought of mentioning it.

File: frontend/src/lib/utils.tsx [hunk 1]
```diff
@@ -1059,7 +1059,7 @@ export function identifierToHuman(identifier: string | number, caseType: 'senten
 }
 
 export function parseGithubRepoURL(url: string): Record<string, string> {
-    const match = url.match(/^https?:\/\/(?:www\.)?github\.com\/([A-Za-z0-9_.-]+)\/([A-Za-z0-9_.-]+)\/?$/)
+    const match = url.match(/^https?:\/\/(?:www\.)?github\.com\/([A-Za-z0-9_.-]+)\/([A-Za-z0-9_.-]+)\/?/)
     if (!match) {
         throw new Error('Must be in the format: https://github.com/user/repo')
```
Comment (yakkomajuri) [inline frontend/src/lib/utils.tsx:1064]: This error is now incorrect I believe

File: posthog/models/plugin.py [hunk 2]
```diff
@@ -71,11 +71,12 @@ def update_validated_data_from_url(validated_data: Dict[str, Any], url: str) ->
     else:
         parsed_url = parse_url(url, get_latest_if_none=True)
         if parsed_url:
-            validated_data["url"] = parsed_url["root_url"]
+            validated_data["url"] = url
```
Comment (yakkomajuri) [inline posthog/models/plugin.py:74]: will this change affect updates of existing plugins?

File: posthog/plugins/test/plugin_archives.py [hunk 3]
```diff
@@ -22,6 +22,11 @@
...
 )
 
+HELLO_WORLD_PLUGIN_GITHUB_SUBDIR_ZIP = (
```
Comment (yakkomajuri) [inline posthog/plugins/test/plugin_archives.py:25]: lol I hate these tests - something to consider in the future

**Review by yakkomajuri — COMMENTED** [review]
This is looking good!

Thanks for the tests - they do cover things well. One thing I'd love though is a test for upgrading plugins under the various structures. e.g. upgrading a plugin with a path, and upgrading a plugin with the old URL (root). Maybe there's some old upgrade tests I'm not aware of, so in that case just refer me to them!

File: frontend/src/lib/utils.tsx (hunk 1 above)
Comment (pjhul) [reply frontend/src/lib/utils.tsx:1064]: Ah yup, I'll update the error

File: posthog/plugins/test/plugin_archives.py (hunk 3 above)
Comment (pjhul) [reply posthog/plugins/test/plugin_archives.py:25]: I know 😆 Trying to edit this file was a battle. Think these tests are still valuable as I found a couple errors in my implementation when running them, but there's probably a more efficient way of at least storing the test data.

**Comment by pjhul** [comment]
Thanks for taking a look!

@yakkomajuri It sounds like it would be useful to have some more tests around updating plugins and making sure this works as expected - I can add these later today. Afaik right now we don't really have many tests for plugin updates. As I mentioned above, with this update in its current state we still wouldn't be fetching updates correctly (if you passed a specific branch we couldn't fetch the latest commit), so I think it makes sense to go through and fix this also as part of this patch. 👍 

@mariusandra I 100% agree that NPM packages are the more robust approach and most likely will be the one I take once everything is setup. I chose to start with GitHub URLs for two main (short-term) reasons. 
1. Simplicity and ease of updating apps - Since I'm expecting a decent number of apps to need fixes once I start testing them all, it'll be really nice to just push and test things out without having to wait for packages to build/publish.
2. Being able to load a plugin from a branch other than main _(This claims to be supported right now but doesn't really work in practice)_ - For testing updates in the short-term (before things are even in a monorepo) this will be _super_ nice to have.

Once all the current apps are in a good spot and I've set up testing, I think then setting up a real CI/CD process to publish all the apps to NPM is the way to go.

**Comment by pjhul** [comment]
Closing as this PR seems to be in Snapshot purgatory. Will reopen soon 👍

**Pull request closed without merging**